### ISSUE TITLE
Sorted TFrecords writer for beam

### ISSUE THREAD

**Issue by pierrot0** [tensorflow/datasets#677]
tfrecords_writer will eventually replace writer in file_format_adapater module (See S3 experiment, used in imagenet, mnist and cifar).
tfrecords_writer writes records in order (sorted by key), which enables the slicing API to be stable.

beam writer still writes round-robin fashion. We should have something equivalent to tfrecords_writer, to order a pcollection by key and write to tfrecords in that order (if possible, with same shards distribution as in tfrecords_writer.

We most probably want to follow the same pattern as for tfrecords_writer:
 1- all records are written to temporary files. Total size and number of examples are recorded.
 2- Number of shards and distribution is computed.
 3- Records stored in temporary files are re-distributes to shards, while keeping order.

2 is the same as it is now. 1 and 3 should be using beam.
The number of shards for temporary storage will most probably have to be specified before the number of records or total size are known. Number of shard for temporary storage could be set to an arbitrary high number of care is taken not to open all files simultaneously when running on local machine.

**Comment by pierrot0** [comment]
Let me add a few more details here.

We already have a pcollection given by the BeamBuilder. The BeamBuilder needs to be changed to return a pcollection of (key, record) tuples instead of record instances.

We need to count the number of elements and total size to compute the number of desired shards. We could be done with something like that:

```py
from apache_beam import typehints
from apache_beam.transforms import core
from apache_beam.transforms import ptransform


@typehints.with_input_types(str)
@typehints.with_output_types(typehints.Tuple[int, int])
class _CountAndSizeCombineFn(core.CombineFn):
  """CombineFn for computing PCollection length and size (bytes)."""

  def create_accumulator(self):
    return (0, 0)  # (Count, size)

  def add_input(self, accumulator, element):
    count, size = accumulator
    count += 1
    size += len(element)
    return count, size

  def add_inputs(self, accumulator, elements):
    count, size = accumulator
    count += len(elements)
    size += sum(len(element) for element in elements)
    return count, size

  def merge_accumulators(self, accumulators):
    count = sum(acc[0] for acc in accumulators)
    size = sum(acc[1] for acc in accumulators)
    return count, size

  def extract_output(self, accumulator):
    return accumulator


# To get the number of elements and size (bytes) of pcollection.
@ptransform.ptransform_fn
def get_len_size(pcollection):
  return pcollection | 'GetLenSize' >> core.CombineGlobally(
      _CountAndSizeCombineFn())
```

We need to compute the hash(key) and maybe (see bellow) order the pcollection using that:

```py
from tensorflow_datasets.core import hashing
from tensorflow_datasets.core import lazy_imports

def shuffle_pcollection(pcollection, split_name):
  """Shuffle PCollection of (key, obj) tuples and, transforming key into hash(key) (int128)."""
  beam = lazy_imports.lazy_imports.apache_beam
  hasher = hashing.Hasher(split_name)
  return (
      pcollection
      | 'Hash key' >> beam.Map(lambda t: (hasher.hash_key(t[0]), t[1]))
      | 'Shuffle' >> beam.ReshufflePerKey()
      )
```

And then we need to dump the pcollection with the sharding we want. This probably excludes the beam tfrecord writer, unless it is given a single shard at a time. Something like:

```py
def _dump_shard(shard_number, records):
  """Dumps given records into shard #shard_number."""
  # at this point records should be small enough to fit in-memory.
  # We might want to check beam behaves correctly when running locally, and doesn't try
  # to run 10 instances of this function in parallel, if each function wants to load 1-2 GiB of data in ram.
  # We might want to re-use code from tfrecord_writer here, or could use beam tfrecord writer directly.

def dump_pcollection(pcollection):
 # TODO: check order is maintained (probably not). If not: don't do the ReshufflePerKey above,
 # Just group by key on shard number, then sort on hash(key) then dump.
 return (pcollection
 | 'Compute shard' >> beam.Map(lambda t: (get_shard(t[0]), t[1]))
 | 'Group by shard' >> beam.GroupByKey()
 | 'Dump shards' >> beam.Map(_dump_shard)
)
```

**Comment by pierrot0** [comment]
Adding Recep. Guys, feel free to tackle this together. Again, this is an interesting issue for you to solve. Also, it is quite needed.

Thanks!

**Comment by ChanchalKumarMaji** [comment]
```
# We might want to check beam behaves correctly when running locally, and doesn't try
# to run 10 instances of this function in parallel, if each function wants to load 1-2 GiB of data in ram.
```
can you please tell me how should I ensure this ?

**Comment by pierrot0** [comment]
Just add some log statements at beginning and end of function, you'll see if they overlap when running locally.

**Comment by ChanchalKumarMaji** [comment]
Some Doubts:
* The `beam_dataset` generates a `pcollection` of (key, record) tuples,
* Then we transform the `pcollection` to (_key, record) tuples where `_key` is 128 bits, then we do a ReshufflePerKey to sort the `pcollection`,
* Then we use the `example_serializer` to convert the each record in pcollection to a format suitable for writing,
* Then we transform the `pcollection` to contain only serialized `record`s and remove the `_key` as the `_key` is no longer needed,
* Then we use `_CountAndSizeCombineFn` to count the number of elements in the pcollection(which is equal to the number of examples in the particular split) and the total size of the split,
* Then we calculate the number of shards needed from the size of the split obtained above,
* Then we follow the code as in `tfrecord_writer`.

Is this process OK ?

**Comment by pierrot0** [comment]
(sorry for lag, was on vacation)
Sounds good Chanchal!
Please let us know how it goes.
